File: driver.m
function []=driver(opt,problem,seedNo,filename)
    tic
    rng(seedNo)
    % Solve the multimodal optimization problem given the problem and optimization options  
    process=OptimProcess(opt,problem); % create the object process
    archive=Archive(problem); % create the object archive
    while 1
        restart=Restart(process,opt,problem); % create the object restart (restart-specific information)
        subpop=restart.initialize_subpop(archive,process,opt,problem); % initialize the subpopulation
        if problem.isDynamic && process.dynamics.recInd<=numel(process.dynamics.recStepSize)
            subpop.center=process.dynamics.recCenter(process.dynamics.recInd,:);
            subpop.center=min([max([subpop.center;problem.lowBound]);problem.upBound]);
            subpop.mutProfile.smean=process.dynamics.recStepSize(process.dynamics.recInd);
            process.dynamics.recInd=process.dynamics.recInd+1;
        end
        restart.run_one_restart(subpop,archive,process,opt,problem); % evolve the subpopulation until convergence 
        if (restart.terminationFlag==-10) % evaluation budget finished
           disp('inside main_par')
           [problem.numCallF ]
           process.dynamics.update(archive,opt,problem);
           archive.usedEvalHist=[archive.usedEvalHist 0];

           process.update_due_to_change(restart,archive,opt,problem);  


           break  
        elseif (restart.terminationFlag==-5) % problem changed           
            process.dynamics.update(archive,opt,problem);
            archive.usedEvalHist=[archive.usedEvalHist 0];
            process.update_due_to_change(restart,archive,opt,problem);  
            process.reset_static(opt, problem);
            archive=Archive(problem); % create new object archive
        else        
            archive.update(restart,process,opt,problem); % update archive
            process.update(restart,archive,opt,problem); % update process
            disp(['timestep=' num2str(process.dynamics.currentTimeStep) ', restartNo = ',   num2str(process.restartNo-1), ', usedEval = ' num2str(process.usedEvalTillRestart) , '=' , num2str(problem.numCallF), ' (' num2str(problem.numCallF/problem.maxEval*100)  '%), archiveSize = ' num2str(archive.size) ' ,bestVal=' num2str(restart.bestVal)] ) % print optimization progress summary after each restart
        end
    end
    if strcmp(problem.suite,'GMPB')
        offErr=problem.extProb.CurrentError;
        save(filename,'offErr','-v7.3')
    else
        save(filename,'problem','process','archive','-v7.3')
    end
    movefile(filename,'result')
end


   

File: driver.py
"""
Driver function for solving multimodal optimization problems.

Written by Ali Ahrari (aliahrari1983@gmail.com)
Translated to Python from MATLAB
"""

import numpy as np
import time
import pickle
from pathlib import Path
from optim_process import OptimProcess
from archive import Archive
from restart import Restart


def driver(opt, problem, seedNo, filename):
    """
    Solve the multimodal optimization problem given the problem and optimization options.
    
    Args:
        opt: Optimization options
        problem: Problem instance
        seedNo: Random seed
        filename: Output filename for saving results
    """
    start_time = time.time()
    np.random.seed(seedNo)
    
    process = OptimProcess(opt, problem)
    archive = Archive(problem)
    
    while True:
        restart = Restart(process, opt, problem)
        subpop = restart.initialize_subpop(archive, process, opt, problem)
        
        # Use predicted centers/step sizes for dynamic problems
        if problem.isDynamic and process.dynamics.recInd <= len(process.dynamics.recStepSize):
            subpop.center = process.dynamics.recCenter[process.dynamics.recInd - 1, :]
            subpop.center = np.minimum(np.maximum(subpop.center, problem.lowBound), problem.upBound)
            subpop.mutProfile.smean = process.dynamics.recStepSize[process.dynamics.recInd - 1]
            process.dynamics.recInd += 1
        
        restart.run_one_restart(subpop, archive, process, opt, problem)
        
        if restart.terminationFlag == -10:  # Evaluation budget finished
            print('inside driver')
            print(f'[{problem.numCallF}]')
            process.dynamics.update(archive, opt, problem)
            archive.usedEvalHist = np.append(archive.usedEvalHist, 0)
            process.update_due_to_change(restart, archive, opt, problem)
            break
        elif restart.terminationFlag == -5:  # Problem changed
            process.dynamics.update(archive, opt, problem)
            archive.usedEvalHist = np.append(archive.usedEvalHist, 0)
            process.update_due_to_change(restart, archive, opt, problem)
            process.reset_static(opt, problem)
            archive = Archive(problem)  # Create new archive
        else:
            archive.update(restart, process, opt, problem)
            process.update(restart, archive, opt, problem)
            print(f'timestep={process.dynamics.currentTimeStep}, restartNo={process.restartNo - 1}, '
                  f'usedEval={process.usedEvalTillRestart}={problem.numCallF} '
                  f'({problem.numCallF / problem.maxEval * 100:.2f}%), '
                  f'archiveSize={archive.size}, bestVal={restart.bestVal}')
    
    # Save results
    result_dir = Path('result')
    result_dir.mkdir(exist_ok=True)
    
    if problem.suite == 'GMPB':
        offErr = problem.extProb.CurrentError
        with open(filename, 'wb') as f:
            pickle.dump({'offErr': offErr}, f)
    else:
        with open(filename, 'wb') as f:
            pickle.dump({
                'problem': problem,
                'process': process,
                'archive': archive
            }, f)
    
    # Move file to result directory
    Path(filename).replace(result_dir / Path(filename).name)
